### ISSUE TITLE
Sending messages without key gives an error

### ISSUE THREAD

**Issue by kpala** [tulios/kafkajs#193]
To my understanding, `key` should be optional when sending messages. However, the following gives the error `"string" must be a string, Buffer, or ArrayBuffer`:

```javascript
producer.send({
  topic: 'topic-name',
  messages: [{
    value: 'Hello world' 
  }]
})
```

Providing `null` as the key seems to be a workaround for now:

```javascript
producer.send({
  topic: 'topic-name',
  messages: [{
    key: null,
    value: 'Hello world' 
  }]
})
```

**Comment by tulios** [comment]
You are right; it might be a regression.

**Comment by ianwsperber** [comment]
@tulios I accidentally referenced this ticket in my PR (#208) for #197, going to re-open this since I assume it was closed due to an error in my part ! 😬

**Comment by Nevon** [comment]
I can't reproduce this issue. There are several tests that are producing messages without any `key`. For example, https://github.com/tulios/kafkajs/blob/a12a00e5222028ee03f4d8845ec95c4e2f9c7852/src/producer/index.spec.js#L79-L82

Are you still able to reproduce this issue, and if so, do you have any more information?

**Comment by kpala** [comment]
I'm still seeing this issue. Here's the minimal code to reproduce:

```
const client = new kafka.Kafka({
    brokers: ['127.0.0.1:9092'],
    clientId: 'test-clientId'
});

const producer = client.producer();
await producer.connect();
await producer.send({
    messages: [{ value: 'Hello world' }],
    topic: 'test-topic'
});

await producer.disconnect();
```

Here's my test output:

```
test
{"level":"ERROR","timestamp":"2019-01-17T13:32:59.334Z","logger":"kafkajs","message":"[Producer] The \"string\" argument must be one of type string, Buffer, or ArrayBuffer. Received type undefined","retryCount":0,"retryTime":283}
  1) send without key


0 passing (8s)
1 failing

1) test
     send without key:
   TypeError [ERR_INVALID_ARG_TYPE]: The "string" argument must be one of type string, Buffer, or ArrayBuffer. Received type undefined
    at Function.byteLength (buffer.js:515:11)
    at Function.sizeOfVarIntBytes (node_modules/kafkajs/src/protocol/encoder.js:48:47)
    at module.exports (node_modules/kafkajs/src/protocol/recordBatch/record/v0/index.js:35:13)
    at messages.map (node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:76:5)
    at Array.map (<anonymous>)
    at .../node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:75:28
    at .../node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:58:34
    at encode (node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:42:35)
    at module.exports (node_modules/kafkajs/src/protocol/request.js:4:25)
    at sendRequest (node_modules/kafkajs/src/network/connection.js:232:36)
    at Connection.send (node_modules/kafkajs/src/network/connection.js:264:59)
    at Broker.produce (node_modules/kafkajs/src/broker/index.js:198:34)
    at brokersWithoutResponse.map (node_modules/kafkajs/src/producer/sendMessages.js:79:41)
    at Array.map (<anonymous>)
    at createProducerRequests (node_modules/kafkajs/src/producer/sendMessages.js:66:37)
```

Works without problems, if I replace `messages: [{ value: 'Hello world' }],` with `messages: [{ key: null, value: 'Hello world' }],`.

**Comment by Nevon** [comment]
What version of kafkajs are you using when seeing that issue?

**Comment by kpala** [comment]
When I reported the issue, it was 1.4.4. Today when running the test above, it's 1.4.6.

**Comment by Nevon** [comment]
I wasn't able to reproduce it because the producer test doesn't use 0.11 features by default, and this issue was only present in Produce V3 requests. Adding that flag to the test made it possible to reproduce.

Got a PR up to fix it, #247. Thanks for reporting!

**Comment by manypozo** [comment]
Hello,
I am having this issue as well.

I am using kafkajs `1.12.0`
brokers are on version `2.4.0`
node 12  (producer docker image from `node:12-alpine`)


My producer is slightly different to the one state in the issue, since I am using schema-registry as well

here is a sample:
```
  async sendRecords(records, topicName, schemaName) {
    let messages = []

    if (schemaName in schemaMap) {
      const schemaID = schemaMap[schemaName].id
      logger.debug(records.length + ' records to send to topic ' + topicName + ' | avroSchema=' + schemaName + ' | SchemaID=' + schemaID)

      const encodedRecordsArray = records.map(async (record) => {
        const avroValue = await registry.encode(schemaID, record)
        return {
          key: null,
          value: avroValue,
          headers: {
            partition: record.type === 'custom' ? record.bundle_id : record.name,
          },
        }
      })

      messages = await Promise.all(encodedRecordsArray)
    } else {
      const encodedRecordsArray = records.map(async (record) => {
        return {
          key: null,
          value: record,
        }
      })

      messages = await Promise.all(encodedRecordsArray)
    }

    // Send records to producer
    await producer.send({
      topic: topicName,
      messages: messages,
      acks: 1,
      compression: CompressionTypes.None,
    })
      .then(() => {
        logger.debug(`${messages.length} records successfully sent to ${topicName}`)
      })
      .catch(err => {
        logger.error(`[kafka/producer] Failed sending ${messages.length} messages to topic ${topicName}`, err)
        logger.debug(`[kafka/producer] Failed sending ${JSON.stringify(records)}`)
      })
  },
```

with output:
```
{"level":"ERROR","timestamp":"2020-09-08T17:24:30.567Z","logger":"kafkajs","message":"[Producer] The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array","retryCount":0,"retryTime":359}
{"level":"ERROR","timestamp":"2020-09-08T17:24:30.573Z","logger":"kafkajs","message":"[Producer] The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array","retryCount":0,"retryTime":286}
{"level":"ERROR","timestamp":"2020-09-08T17:24:30.582Z","logger":"kafkajs","message":"[Producer] The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array","retryCount":0,"retryTime":300}
{"level":"ERROR","timestamp":"2020-09-08T17:24:30.585Z","logger":"kafkajs","message":"[Producer] The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array","retryCount":0,"retryTime":348}
{"app":"vs-data-api","time":"2020-09-08T17:24:30.589Z","level":"error","msg":"[kafka/producer] Failed sending 444 messages to topic my_topic_1","err":"TypeError [ERR_INVALID_ARG_TYPE]: The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array\n    at Function.byteLength (buffer.js:726:11)\n    at Function.sizeOfVarIntBytes (/home/node/app/node_modules/kafkajs/src/protocol/encoder.js:48:46)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/recordBatch/record/v0/index.js:36:13)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:96:5\n    at Array.map (<anonymous>)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:95:28\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:70:13\n    at encode (/home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:46:15)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/request.js:4:25)\n    at sendRequest (/home/node/app/node_modules/kafkajs/src/network/connection.js:276:36)"}
{"app":"vs-data-api","time":"2020-09-08T17:24:30.589Z","level":"error","msg":"[kafka/producer] Failed sending 605 messages to topic my_topic_2","err":"TypeError [ERR_INVALID_ARG_TYPE]: The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array\n    at Function.byteLength (buffer.js:726:11)\n    at Function.sizeOfVarIntBytes (/home/node/app/node_modules/kafkajs/src/protocol/encoder.js:48:46)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/recordBatch/record/v0/index.js:36:13)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:96:5\n    at Array.map (<anonymous>)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:95:28\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:70:13\n    at encode (/home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:46:15)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/request.js:4:25)\n    at sendRequest (/home/node/app/node_modules/kafkajs/src/network/connection.js:276:36)"}
{"app":"vs-data-api","time":"2020-09-08T17:24:30.589Z","level":"error","msg":"[kafka/producer] Failed sending 206 messages to topic my_topic_3","err":"TypeError [ERR_INVALID_ARG_TYPE]: The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array\n    at Function.byteLength (buffer.js:726:11)\n    at Function.sizeOfVarIntBytes (/home/node/app/node_modules/kafkajs/src/protocol/encoder.js:48:46)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/recordBatch/record/v0/index.js:36:13)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:96:5\n    at Array.map (<anonymous>)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:95:28\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:70:13\n    at encode (/home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:46:15)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/request.js:4:25)\n    at sendRequest (/home/node/app/node_modules/kafkajs/src/network/connection.js:276:36)"}
{"app":"vs-data-api","time":"2020-09-08T17:24:30.589Z","level":"error","msg":"[kafka/producer] Failed sending 1 messages to topic my_topic_4","err":"TypeError [ERR_INVALID_ARG_TYPE]: The \"string\" argument must be of type string or an instance of Buffer or ArrayBuffer. Received an instance of Array\n    at Function.byteLength (buffer.js:726:11)\n    at Function.sizeOfVarIntBytes (/home/node/app/node_modules/kafkajs/src/protocol/encoder.js:48:46)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/recordBatch/record/v0/index.js:36:13)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:96:5\n    at Array.map (<anonymous>)\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:95:28\n    at /home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:70:13\n    at encode (/home/node/app/node_modules/kafkajs/src/protocol/requests/produce/v3/request.js:46:15)\n    at module.exports (/home/node/app/node_modules/kafkajs/src/protocol/request.js:4:25)\n    at sendRequest (/home/node/app/node_modules/kafkajs/src/network/connection.js:276:36)"}
```


maybe I am doing something wrong.

Thank you!
